### PR TITLE
Implement exclude file regex for projects and workspace

### LINKED ISSUE #5508: Test framework info bar detects excluded files for "*test*.py" pattern.
The info bars for configure test framework checks the project or workspace for a file named "*test.py*". Currently, it checks the entire project or workspace folder and subfolders, whereas it should not check for files which are not included in the project or workspace. For example, if there is a virtual environment folder in a workspace, all files in that folder should not be checked for "*test*.py" file name. If a file is not included as part of a project (in .pyproj definition), then it should not be checked. 

Related to #5496

### PR DESCRIPTION
**Pull request by RaymonGulati1**
Fix #5508 

Regex will handle most cases correctly, but it might encounter some edge cases it cannot correctly parse. For example, "test.test.py" or "test.txtfpp" result in false positives.

These cases will be correctly parsed: 
image

### CODE REVIEW

**Comment by RaymonGulati1** [comment]
Still writing tests

File: Python/Product/PythonTools/PythonTools/Project/ConfigureTestFrameworkInfoBar.cs [hunk 1]
```diff
@@ -281,6 +278,15 @@ internal sealed class TestFrameworkProjectInfoBar : TestFrameworkInfoBar {
             Project.SetProjectProperty(propertyName, propertyValue);
             return Task.CompletedTask;
         }
+
+        protected override bool PythonTestFileFound(Predicate<string> fileFilter) {
+            return Project.AllDescendants
+                .Where(x => (x is PythonFileNode || x is PythonNonCodeFileNode))
```
Comment (huguesv) [inline Python/Product/PythonTools/PythonTools/Project/ConfigureTestFrameworkInfoBar.cs]: did you test that this correctly returns false for files that are excluded from project?

File: Python/Product/VSInterpreters/Interpreter/PythonWorkspaceContext.cs [hunk 2]
```diff
@@ -152,6 +152,29 @@ class PythonWorkspaceContext : IPythonWorkspaceContext {
...
+                ) {
+                    foreach (var file in Directory
+                                .EnumerateFiles(topLevelDirectory, "*", SearchOption.AllDirectories)
+                                .Where(x => predicate(x))
+                    ) {
+                        yield return file;
+                    }
+                }
+            }
+        }
```
Comment (huguesv) [inline Python/Product/VSInterpreters/Interpreter/PythonWorkspaceContext.cs]: missing a blank line

File: Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs [hunk 3]
```diff
@@ -176,7 +177,8 @@ class TestContainerDiscovererWorkspace : ITestContainerDiscoverer, IDisposable {
                 _testFilesUpdateWatcher.AddDirectoryWatch(workspace.Location);
                 oldWatcher?.Dispose();
 
-                foreach (var file in GetTestContainerFiles(projInfo)) {
+                Predicate<string> testFileFilter = (x) => Regex.IsMatch(PythonConstants.TestPatternFileNameRegex, x);
+                foreach (var file in _workspaceContextProvider.Workspace.EnumerateUserFiles(testFileFilter)) {
                     projInfo.AddTestContainer(this, file);
```
[the remark this answers is not in the document]
Comment (bschnurr) [reply Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs]: To add .txt file support we need to update AddTestContainer(..)

File: Python/Product/PythonTools/PythonConstants.cs [hunk 4]
```diff
@@ -134,6 +134,7 @@ static class PythonConstants {
         public const string DefaultUnitTestRootDirectory = ".";
         public const string UnitTestPatternSetting = "UnitTestPattern";
         public const string DefaultUnitTestPattern = "test*.py";
+        public const string TestPatternFileNameRegex = @"((^test.*)|(^.*_test))\.(py|txt)";
```
[the remark this answers is not in the document]
Comment (RaymonGulati1) [reply Python/Product/PythonTools/PythonConstants.cs]: Done

File: Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (RaymonGulati1) [reply Python/Product/TestAdapter/TestContainerDiscovererWorkspace.cs]: Changed

File: Python/Product/PythonTools/PythonTools/Project/ConfigureTestFrameworkInfoBar.cs (hunk 1 above)
Comment (RaymonGulati1) [reply Python/Product/PythonTools/PythonTools/Project/ConfigureTestFrameworkInfoBar.cs]: Yes
